ui/components/component-library/text-field-base/text-field-base.js

```diff
@@ -0,0 +1,285 @@
+import React, { useState, useRef, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+
+import {
+  DISPLAY,
+  COLORS,
+  SIZES,
+  ALIGN_ITEMS,
+  TEXT,
+} from '../../../helpers/constants/design-system';
+
+import Box from '../../ui/box';
+
+import { Text } from '../text';
+
+import { TEXT_FIELD_SIZES } from './text-field-base.constants';
+
+export const TextFieldBase = React.forwardRef(
+  (
+    {
+      autoFocus,
+      className,
+      defaultValue,
+      disabled,
+      error,
+      id,
+      inputProps,
+      inputRef,
+      leftAccessory,
+      rightAccessory,
+      maxLength,
+      name,
+      onBlur = () => {},
+      onChange = () => {},
+      onClick = () => {},
+      onFocus = () => {},
+      onKeyDown = () => {},
+      onKeyUp = () => {},
+      placeholder,
+      readOnly,
+      required,
+      size = SIZES.MD,
+      type = 'text',
+      truncate,
+      value: inputValue,
+      ...props
+    },
+    ref,
+  ) => {
+    const internalInputRef = useRef(null);
+
+    const [value, setValue] = useState(defaultValue || inputValue || '');
+    const [focused, setFocused] = useState(false);
+
+    useEffect(() => {
+      setValue(inputValue);
+    }, [inputValue]);
+
+    useEffect(() => {
+      // The blur won't fire when the disabled state is set on a focused input.
+      // We need to set the focused state manually.
+      if (disabled) {
+        setFocused(false);
+      }
+    }, [disabled]);
+
+    const handleClick = (event) => {
+      const { current } = internalInputRef;
+
+      if (current) {
+        current.focus();
+        setFocused(true);
+      }
+
+      if (onClick) {
+        onClick(event);
+      }
+    };
+
+    const handleChange = (event) => {
+      // setState handles the use case of an uncontrolled component
+      setValue(event.target.value);
+      // onChange will be used for controlled components
+      onChange(event);
+    };
+
+    const handleFocus = (event) => {
+      setFocused(true);
+      onFocus(event);
+    };
+
+    const handleBlur = (event) => {
+      setFocused(false);
+      onBlur(event);
+    };
+
+    const handleKeyDown = (event) => {
+      onKeyDown(event);
+    };
+
+    const handleKeyUp = (event) => {
+      onKeyUp(event);
+    };
+
+    const handleInputRef = (ref) => {
+      internalInputRef.current = ref;
+      if (inputRef && inputRef.current !== undefined) {
+        inputRef.current = ref;
+      } else if (typeof inputRef === 'function') {
+        inputRef(ref);
+      }
+    };
```

Comment:
Need to figure out my uncontrolled > controlled issue

<img width="1425" alt="Screen Shot 2022-09-30 at 11 57 58 AM" src="https://user-images.githubusercontent.com/8112138/193338479-a996c2ff-3271-42b5-be18-c5ced9ba82fa.png">